radix-engine-stores/src/hash_tree/tree_change.rs

```diff
@@ -0,0 +1,169 @@
+use crate::hash_tree::{BatchChange, HashChange, IdHashChange};
+use radix_engine_common::crypto::Hash;
+use radix_engine_store_interface::interface::{
+    DbNodeKey, DbPartitionKey, DbPartitionNum, DbSortKey,
+};
+use utils::prelude::{index_map_new, IndexMap};
+use utils::rust::vec::Vec;
+
+/// A canonical form of "change to an entire hash tree".
+/// Use `From<Vec<HashChange>>` implementation to canonicalize a sequence of changes.
+pub enum TreeChange {
+    /// A "delta" - the only supported type of a global change.
+    /// Other global changes are conceivable (e.g. "clear entire tree"), but not needed.
+    Delta {
+        node_changes: Vec<(DbNodeKey, NodeTierChange)>,
```

Comment:
> repeat `DbNodeKey` in different orders would represent different updates

repeat `DbNodeKey` would be an invalid state of a `TreeChange`.
As in: the `TreeChangeBuilder`'s impl does not allow you to build such instance of a `TreeChange`.
(ofc you can still build it via hacks, but that can be said about any broken invariant)

After the build, the `TreeChange`'s impl does not require Map-like access for anything.

>  does it make better sense to have this be a `Map<DbNodeKey, NodeTierChange>`

It is a bit like this hypothetical `Timeseries` struct:
```
struct Timeseries<T> {
  datapoints: Vec<(Instant, T)>,
}
```
Here one could say "you better make `datapoints` a `BTreeMap<Instant, E>` since they gotta be sorted at all times".
Yet, metrics libraries typically use Lists/Arrays/Vectors there.
They of course still keep invariants: e.g. `record_next(Instant at, T value)` checks that `at >= datapoints.last.0`.
With such guarantees, the `Timeseries` struct above can very well be called "a canonical representation" (where "canonicalization" is `sort()`).
_[I am sorry for a bit too obvious example, but 90% of my background is metrics collection]_

Reply:
Gotcha makes sense, totally fine with this datastructure. I think just worth a comment regarding this as normally `datapoints` in your example would be hidden as an implementation detail. In this case, as part of an `enum` its forced to be public and knowing what the invariants to the structure should be more explicit.

I think when talking about "changes", you are usually talking about a series of changes in which order is important. In this particular case, it doesn't follow that construct which is why it was confusing to me at first reading.